src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/CombinationController.php

```diff
@@ -124,48 +125,65 @@ public function bulkEditFormAction(int $productId): Response
      *
      * @param Request $request
      * @param int $productId
-     * @param int $combinationId
      *
      * @return JsonResponse
      */
-    public function bulkEditAction(Request $request, int $productId, int $combinationId): JsonResponse
+    public function bulkEditAction(Request $request, int $productId): JsonResponse
     {
-        try {
-            // PATCH request is required to avoid disabled fields to be forced with null values
-            $bulkCombinationForm = $this->getBulkCombinationFormBuilder()->getFormFor($combinationId, [], [
-                'method' => Request::METHOD_PATCH,
-                'product_id' => $productId,
-            ]);
-        } catch (CombinationNotFoundException $e) {
-            return $this->returnErrorJsonResponse(
-                ['error' => $this->getErrorMessageForException($e, $this->getErrorMessages($e))],
-                Response::HTTP_NOT_FOUND
-            );
+        $combinationIds = $request->request->get('combinationIds');
+        if (!$combinationIds) {
+            return $this->json([
+                'fatal' => true,
+                'errors' => $this->getFallbackErrorMessage('', 0, 'Missing combinationIds in request body'),
+            ], Response::HTTP_BAD_REQUEST);
         }
 
-        try {
-            $bulkCombinationForm->handleRequest($request);
-            $result = $this->getBulkCombinationFormHandler()->handleFor($combinationId, $bulkCombinationForm);
-
-            if (!$result->isSubmitted()) {
-                return $this->json(['errors' => [
-                    'form' => [
-                        $this->trans('No submitted data.', 'Admin.Notifications.Error'),
-                    ],
-                ]], Response::HTTP_BAD_REQUEST);
+        $combinationIds = json_decode($combinationIds);
+        $errors = [];
+        foreach ($combinationIds as $combinationId) {
+            try {
+                // PATCH request is required to avoid disabled fields to be forced with null values
+                $bulkCombinationForm = $this->getBulkCombinationFormBuilder()->getFormFor($combinationId, [], [
+                    'method' => Request::METHOD_PATCH,
+                    'product_id' => $productId,
+                ]);
+            } catch (CombinationNotFoundException $e) {
+                $errors[] = $this->getErrorMessageForException($e, $this->getErrorMessages($e));
+                continue;
             }
 
-            if ($result->isValid()) {
-                return $this->json([]);
+            try {
+                $bulkCombinationForm->handleRequest($request);
+                $result = $this->getBulkCombinationFormHandler()->handleFor($combinationId, $bulkCombinationForm);
```

Comment:
```suggestion
                $result = $this->getBulkCombinationFormHandler()->handleFor($combinationIds, $bulkCombinationForm);
```

Then the trick would be here, the `FormDataHandlerInterface:update` is not strictly typed because it can be used with many different types (int, string, VOs, ...) so we can take advantage of this and provide the list of IDs here.

This way the form is only built once, it handles the request once, and even the form handle is only called once. The loop will actually be performed in the `BulkCombinationFormDataHandler:update($combinationIds, array $data): void` method

Reply:
but this way we will break the contract of interface (its like hacking :smile: )